java/src/main/java/ai/onnxruntime/OrtSession.java

```diff
@@ -199,7 +194,7 @@ public long getNumOutputs() {
    *     there are zero or too many inputs.
    */
   public Result run(Map<String, OnnxTensor> inputs) throws OrtException {
-    return run(inputs, outputNamesArray, null);
+    return run(inputs, outputNamesArray, defaultRunOptions);
```

Comment:
I don't know if the RunOptions instance can be shared between concurrent runs. @pranavsharma Could someone from MS tell us about the thread safety and shareability of these classes?

Reply:
I don't know a lot about CPP thread safety, but i think these are some reads and writes of that terminate flag: 
https://github.com/microsoft/onnxruntime/blob/0638565fe0d6149d1512ec3e63512e6ea13743a3/include/onnxruntime/core/framework/run_options.h#L23
https://github.com/microsoft/onnxruntime/blob/7c89f38a346bcfc4494884d8278ba838495e9575/onnxruntime/core/framework/run_options.cc#L48
https://github.com/microsoft/onnxruntime/blob/274e6b41535f79808e2c6d52406851479ab6dca0/onnxruntime/core/framework/parallel_executor.cc#L129

In this case the terminate flag would be set in another thread from the model execution.

Reply:
Yeah, I'm fairly confident the terminate functionality would be fine, but my worry is if the C++ devs consider this to be a class that is only bound to a single call to run, then they might refactor things in a way that break us without a C API update. Which is why I'd like someone from the core team to check (or tell us where we could find that out from the docs).

Reply:
Given it's a small object, with presumably low overhead (apart from the JNI call to create it) it might be worth manufacturing one each time, as that would avoid this penalty at the cost of slightly increasing latency (for the JNI) and a marginal increase in garbage (for the Java object).

Reply:
Ah, I see you did that in the latest change.

Reply:
What do the maintainers think? Is it possible to setTerminate in one thread and safely stop execution in another thread?

Reply:
All Run() interfaces in Session object take RunOptions by const ptr/reference. The input parameter is not modified.
The references above has to do either with the interfaces that allow you change the default values of RunOptions before running inference.
The last reference has to do with the parallel executor member and not with RunOPtions. It it safe to say that it has nothing to do with Java wrappers.

---
In reply to: [453974752](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r453974752) [](ancestors = 453974752)